core/chain_makers.go

```diff
@@ -238,58 +243,79 @@ func GenerateChain(config *params.ChainConfig, parent *types.Block, engine conse
-		Root:        []common.Hash{common.Hash{}, common.Hash{}, common.Hash{}},
-		Difficulty:  []*big.Int{big.NewInt(1), big.NewInt(1), big.NewInt(1)},
-		UncleHash:   types.EmptyUncleHash,
-		TxHash:      types.EmptyRootHash,
-		ReceiptHash: types.EmptyRootHash,
-		Bloom:       []types.Bloom{types.Bloom{}, types.Bloom{}, types.Bloom{}},
-		Time:        time,
-		BaseFee:     []*big.Int{baseFee, baseFee, baseFee},
-		GasLimit:    []uint64{0, 0, 0},
-		GasUsed:     []uint64{0, 0, 0},
-		Extra:       [][]byte{[]byte(nil), []byte(nil), []byte(nil)},
+		ParentHash:        make([]common.Hash, 3),
+		Number:            []*big.Int{new(big.Int).SetUint64(0), new(big.Int).SetUint64(0), new(big.Int).SetUint64(0)},
+		Extra:             make([][]byte, 3),
+		Time:              uint64(0),
+		BaseFee:           []*big.Int{baseFee, baseFee, baseFee},
+		GasLimit:          make([]uint64, 3),
+		Coinbase:          make([]common.Address, 3),
+		Difficulty:        make([]*big.Int, 3),
+		NetworkDifficulty: make([]*big.Int, 3),
+		Root:              make([]common.Hash, 3),
+		TxHash:            make([]common.Hash, 3),
+		ReceiptHash:       make([]common.Hash, 3),
+		GasUsed:           make([]uint64, 3),
+		Bloom:             make([]types.Bloom, 3),
+		Location:          chain.Config().Location,
 	}
-	header.GasLimit[types.QuaiNetworkContext] = parent.GasLimit()
 
-	parentHeader := &types.Header{
-		Number:     []*big.Int{big.NewInt(int64(1)), big.NewInt(int64(1)), big.NewInt(int64(1))},
-		Difficulty: []*big.Int{big.NewInt(1), big.NewInt(1), big.NewInt(1)},
-		UncleHash:  types.EmptyUncleHash,
-		Time:       time - 10,
-	}
-
-	if chain.Config().IsLondon(header.Number[types.QuaiNetworkContext]) {
-		if !chain.Config().IsLondon(parent.Number()) {
-			parentGasLimit := parent.GasLimit() * params.ElasticityMultiplier
-			header.GasLimit[types.QuaiNetworkContext] = CalcGasLimit(parentGasLimit, parent.GasUsed(), 0)
-		}
+	// same across orders
+	header.Time = parent.Time() + uint64(10)
+	gasLimit := CalcGasLimit(parent.GasLimit(), parent.GasUsed(), len(parent.Uncles()))
+	header.GasLimit = []uint64{gasLimit, gasLimit, gasLimit}
+
+	switch order {
+	case 2: // Zone
+		// fill ParentHash values
+		header.ParentHash[0] = parent.Header().ParentHash[0]
+		header.ParentHash[1] = parent.Header().ParentHash[1]
+		header.ParentHash[2] = parent.Hash()
+		// fill Number values
+		header.Number[0] = parent.Header().Number[0]
+		header.Number[1] = parent.Header().Number[1]
+		header.Number[2].Add(parent.Header().Number[2], common.Big1)
+		// fill Difficulty values
+		header.Difficulty[0] = parent.Header().Difficulty[0]
+		header.Difficulty[1] = parent.Header().Difficulty[1]
+		header.Difficulty[2] = ethash.CalcDifficulty(config, header.Time, parent.Header(), 2)
+	case 1: // Region
+		// fill ParentHash values
+		header.ParentHash[0] = parent.Header().ParentHash[0]
+		header.ParentHash[1] = parent.Hash()
+		header.ParentHash[2] = header.ParentHash[1]
```

Comment:
I'm not sure this is valid. The zone parent usually wont be the region parent. There will very likely be several zone blocks between each region block.